tsnet/tsnet_test.go

```diff
@@ -44,3 +56,98 @@ func TestListenerPort(t *testing.T) {
 		}
 	}
 }
+
+var verboseNodes = flag.Bool("verbose-nodes", false, "if set, print tsnet.Server logs")
+
+func TestConn(t *testing.T) {
+	derpMap := integration.RunDERPAndSTUN(t, logger.Discard, "127.0.0.1")
+	control := &testcontrol.Server{
+		DERPMap: derpMap,
+	}
+	control.HTTPTestServer = httptest.NewUnstartedServer(control)
+	control.HTTPTestServer.Start()
+	t.Cleanup(func() {
+		control.HTTPTestServer.Close()
+	})
+	controlURL := control.HTTPTestServer.URL
+	t.Logf("testcontrol listening on %s", controlURL)
+
+	s1 := &Server{
+		ControlURL: controlURL,
+		Hostname:   "s1",
+		Store:      new(mem.Store),
+		Ephemeral:  true,
+	}
+	defer s1.Close()
+
+	s2 := &Server{
+		ControlURL: controlURL,
+		Hostname:   "s2",
+		Store:      new(mem.Store),
+		Ephemeral:  true,
+	}
+	defer s2.Close()
+
+	if !*verboseNodes {
+		s1.Logf = logger.Discard
+		s2.Logf = logger.Discard
+	}
+
+	ln, err := s1.Listen("tcp", ":8081")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer ln.Close()
+
+	lc1, err := s1.LocalClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	s1status, err := lc1.Status(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+	s1ip := s1status.TailscaleIPs[0]
+
+	if err := s2.Start(); err != nil {
+		t.Fatal(err)
+	}
+
+	lc2, err := s2.LocalClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dialCtx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+
+	res, err := lc2.Ping(dialCtx, s1ip, tailcfg.PingICMP)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("ping success: %#+v", res)
+
+	w, err := s2.Dial(dialCtx, "tcp", fmt.Sprintf("%s:8081", s1ip))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	r, err := ln.Accept()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	want := "hello"
+	if _, err := io.WriteString(w, want); err != nil {
+		t.Fatal(err)
+	}
+
+	got := make([]byte, len(want))
+	if _, err := io.ReadAtLeast(r, got, len(got)); err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("got: %q", string(got))
```

Comment:
btw %q works on []byte the same as string so this string(got) conversion is unnecessary